### DISCUSSION TITLE
Filter votes by tag

### DISCUSSION THREAD

**Discussion by LiamM32** [julien-boudry/Condorcet#141] [Q&A] [answered]
Hello. I am creating a new pairwise method. I would like a version of the getPairwise function that only counts votes with a specified tag. Is there any way to achieve this functionality with the current API?

**Comment by LiamM32** [comment]
@julien-boudry I have spent considerable time trying to do this now. I have gotten a better understanding of Condorcet's codebase than I expected to. Some functions in _VotesProcess_ appear to have a parameter to filter by tag. But _getPairwise()_ in _ResultsProcess_ doesn't have such an option.

I figure that the best way to do what I want is to replicate the functionality of _getPairwise_, but do it on a filtered array of votes. But I don't know how to do this. I don't understand how _getPairwise_ works. The process to determine the pairwise results on a set of votes appears to be rather complicated. Do you have any advice?

I'm making a specialized application that will never be merged into the core Condorcet project. I'm not particularly against modifying the existing API. If it's straightforward to do this under the existing API, I would be happy to use that. But it may be a good opportunity to make a more flexible function for determining pairwise comparisons.

**Comment by julien-boudry** [comment]
HI @LiamM32,

Nice to hear about your project.

True, it's currently not possible to get Pairwise for a Result filtered by tags.
You're also referring to something pretty ugly in the API and the design, that little tag-filtering hack.

The idea behind Result object is that it contains immutable properties.
<img width="721" alt="image" src="https://github.com/julien-boudry/Condorcet/assets/4020317/b4f9890f-c9e1-4e0a-868a-220dd4659456">

A quick and correct idea could be to add a property containing an immutable Pairwise and could be easy to implement, can do it today :)

Could be :
```php
$r = $e->getResult(methodOptions: ['%tagFilter' => true, 'tags' => ['tag2']]);
pairwise = $r->explicitPairwise;
```

ps: thanks for your sponsorship!

**Reply by julien-boudry** [reply]
Other options, actually existing, it to use de [Election->removeVotesBytags()](https://www.condorcet.io/Docs/ApiReferences/Election%20Class/public%20Election--removeVotesByTags) methods.

Than ask for Pairwise.

**Reply by LiamM32** [reply]
Does parseVotes() get the number of votes ranking A over B? If so, than that is exactly what I want.

Does the getResult() function compute the end rankings for a whole method? That's a good feature, but not exactly what I want.

**Reply by julien-boudry** [reply]
`parseVotes` is like `addVote`but you can parse many votes from a string format (and also the tags).
Here, 2 votes are successfully parsed, so it returns `2`.

Yes, getResult return the result of a whole method, but with the commit 5b11a2e , it contain also the Pairwise (immutable) at this moment.
Normally, it's more of a public API method, but it's possible to use it inside another method, like that from CPO STV: https://github.com/julien-boudry/Condorcet/blob/master/src/Algo/Methods/STV/CPO_STV.php#L266
Combined with the getResult version with tag filter, you can get the filtered pairwise easily.

**Comment by julien-boudry** [comment, marked as answer]
Commit 5b11a2e

Finally, it's:
```php
$r = $e->getResult(methodOptions: ['%tagFilter' => true, '%withTag' => true, 'tags' => ['tag2']]);
$filteredByTagsPairwise = $r->pairwise;
```

And maybe later, iI will split getResult into different and more explicit methods names than this option filter for tags.

It's more a Public API solution than a real internal solution, but it's will work very well internally.

Except that all the Result & Tags systems don't work well with very large elections (only if you manually activate the external data handler), but not a problem before many millions of votes. It's a limitation that must be investigated and documented.

**Comment by LiamM32** [comment]
Perhaps it will be easier for you to understand what kind of function I want if I explain what I'm doing.

The method I am doing requires the _A_ vs _B_ pairwise margin to be determined for votes for a specified tag. It does this for each tag from a list of tags. For each tag, and each pair of candidates, it calculates the following formula: `(($a-$b)*p)^(1/3)`

For each pair of candidates, it takes the sum of outputs of that formula for each tag. The result of that sum will be treated as the pairwise margin for the Schulze (margins) method.

**Reply by julien-boudry** [reply]
The formula seems to be incomplete, and invalid (typo?).

That is possible with my method, in the previous comment, certainly not optimized (in terms of performance) but really practicable.

The only little trick you need: internal methods generally works with candidates keys, not the candidates' names or objects.
So you will need to convert the key to candidates name.

Your implementation of the Schulze variant (the same place as your other Gecko method) could look something like that.

```php
    protected function schulzeVariant(int $i, int $j, Election $election): int
    {
        $score = 0;
        $iName = $election->getCandidateObjectFromKey($i)->getName();
        $jName = $election->getCandidateObjectFromKey($j)->getName();

        foreach (['tag1', 'tag2', 'tag3'] as $tag) {
            $filteredPairwise = $election->getResult(methodOptions: ['%tagFilter' => true, '%withTag' => true, 'tags' => $tag])->pairwise;

            // Do your formula etc.
            $resultOfFormula = $filteredPairwise[$iName]['win'][$jName] - $filteredPairwise[$jName]['win'][$iName];

            $score += $resultOfFormula;
        }

        return $score;
    }
```

And you can see the pairwise format in this test (for A>B>C):
https://github.com/julien-boudry/Condorcet/blob/master/Tests/src/Algo/Methods/PairwiseTest.php#L34

**Reply by LiamM32** [reply]
Yes. The formula had a typo. I edited it. You can look at it again.

**Reply by julien-boudry** [reply]
`$a-$b` is the pairwise margin? What's `p`?

**Reply by LiamM32** [reply]
@julien-boudry `$p` is the population of a country where the voters are located. I am getting it to read from a .json file of national populations. The tags that I'm using are 3-character country codes. This is so that if we don't get a proportional number of voters from each country, votes from underrepresented countries are amplified.

**Reply by julien-boudry** [reply]
Ok :)
Do you have everything with the example above (work only on 4.5 dev branch)?

**Reply by LiamM32** [reply]
I think I have everything. But I think the 11th line there may be a typo. Here is what I suspect you meant to write:
`$resultOfFormula = $filteredPairwise[$iName]['win'][$jName] - $filteredPairwise[$jName]['win'][$iName];`

I don't know how to get the command line program to recognize my new method as an option though. So I haven't tested it.

I also have to find out how to enter tags in votes, as I have never done that.

**Reply by julien-boudry** [reply]
Yes you're right, I edited it.

Methods are registered in the Condorcet.php file, at the beginning. If you don't want to fork the project but build an independent module, this call can be done anywhere, same way, at any time.

For tags, it's easy with the PHP SDK, but not really supported on the command line version (only on display vote option).

But you can ingest votes with tags, do it via ingestion of a [CondorcetElectionFormat](https://github.com/CondorcetVote/CondorcetElectionFormat) (read the spec), as follow: [condorcet.io/2.AsCommandLineApplication/3.ImportFromElectionFormat]( https://www.condorcet.io/2.AsCommandLineApplication/3.ImportFromElectionFormat), it's the only way.

**Reply by julien-boudry** [reply]
Oh, I forgot that you could also do it on the command line by this method.

![image](https://github.com/julien-boudry/Condorcet/assets/4020317/f634048d-ba3a-416d-b266-892c3e286fd2)

**Reply by LiamM32** [reply]
I'm getting an error with the .cvotes file. Here is my attempt to use the [first example](https://github.com/CondorcetVote/CondorcetElectionFormat#valid) you wrote. The stack trace doesn't involve any files that I modified.

```
[liam@Liam-Manjaro-Desktop Condorcet]$ ./bin/condorcet election --import-condorcet-election-format=myTests/Julians-Beautiful-Election.cvotes -lr Schulze Borda Minimax

     _/_/_/                            _/                                            _/   
  _/          _/_/    _/_/_/      _/_/_/    _/_/    _/  _/_/    _/_/_/    _/_/    _/_/_/_/
 _/        _/    _/  _/    _/  _/    _/  _/    _/  _/_/      _/        _/_/_/_/    _/     
_/        _/    _/  _/    _/  _/    _/  _/    _/  _/        _/        _/          _/      
 _/_/_/    _/_/    _/    _/    _/_/_/    _/_/    _/          _/_/_/    _/_/_/      _/_/   

==========================================================================================

Version: 4.5.0-(dev)-g9e977045 || Author: Julien Boudry and contributors || Homepage: https://github.com/julien-boudry/Condorcet

PHP Fatal error:  Uncaught TypeError: CondorcetPHP\Condorcet\Console\Commands\ElectionCommand::parseFromCondorcetElectionFormat(): Argument #1 ($callBack) must be of type Closure, null given, called in /home/liam/src2/Condorcet/src/Console/Commands/ElectionCommand.php on line 356 and defined in /home/liam/src2/Condorcet/src/Console/Commands/ElectionCommand.php:742
Stack trace:
#0 /home/liam/src2/Condorcet/src/Console/Commands/ElectionCommand.php(356): CondorcetPHP\Condorcet\Console\Commands\ElectionCommand->parseFromCondorcetElectionFormat()
#1 /home/liam/src2/Condorcet/src/Console/Commands/ElectionCommand.php(382): CondorcetPHP\Condorcet\Console\Commands\ElectionCommand->importInputsData()
#2 /home/liam/src2/Condorcet/vendor/symfony/console/Command/Command.php(312): CondorcetPHP\Condorcet\Console\Commands\ElectionCommand->execute()
#3 /home/liam/src2/Condorcet/vendor/symfony/console/Application.php(1022): Symfony\Component\Console\Command\Command->run()
#4 /home/liam/src2/Condorcet/vendor/symfony/console/Application.php(314): Symfony\Component\Console\Application->doRunCommand()
#5 /home/liam/src2/Condorcet/vendor/symfony/console/Application.php(168): Symfony\Component\Console\Application->doRun()
#6 /home/liam/src2/Condorcet/src/Console/CondorcetApplication.php(30): Symfony\Component\Console\Application->run()
#7 /home/liam/src2/Condorcet/bin/condorcet.php(32): CondorcetPHP\Condorcet\Console\CondorcetApplication::run()
#8 /home/liam/src2/Condorcet/bin/condorcet(3): require('...')
#9 {main}
  thrown in /home/liam/src2/Condorcet/src/Console/Commands/ElectionCommand.php on line 742
```

**Reply by julien-boudry** [reply]
Just convert it into #147 issue.
